src/llmcompressor/modifiers/awq/base.py

```diff
@@ -0,0 +1,721 @@
+
+            # Q(W * s)
+            for fc in linears2scale:
+                with align_module_device(fc):
+                    fc.weight.mul_(scales_view)
+                    update_offload_parameter(
+                        fc,
+                        "weight",
+                        pseudo_quantize_tensor(
+                            w=fc.weight.data,
+                            symmetric=self.symmetric,
+                            bit_width=self.bits,
+                            group_size=self.group_size,
+                        )[0]
+                        / scales_view,
+                    )
+
+            # W * X
+            int_w_output = self._forward_input_with_kwargs(
+                module=module2inspect, inputs=x, input_kwargs=self.module_kwargs_
+            )
+
+            # compute mean squared error (L2 norm)
+            loss = self._compute_loss(fp16_output, int_w_output, device)
+
+            history.append(loss)
+            if loss < best_error:
+                best_error = loss
+                best_ratio = ratio
+                best_scales = scales.clone()
+            module2inspect.load_state_dict(org_sd)
+
+        if best_ratio == -1:
+            logger.debug(history)
+            raise Exception
+
+        assert (
+            torch.isnan(best_scales).sum() == 0
+        ), f"Nan found in scales: {best_scales}"
+
+        return best_scales.detach().cpu()
+
+    @torch.no_grad()
+    def _compute_loss(
+        self,
+        fp16_output: torch.Tensor,
+        int_w_output: torch.Tensor,
+        device: torch.device,
+    ):
+        loss = 0.0
+        fp16_output_flat = fp16_output.view(-1)
+        int_w_output_flat = int_w_output.view(-1)
+        num_elements = fp16_output_flat.size(0)
+        element_size_bytes = fp16_output.element_size()
+
+        # Calculate chunk size dynamically based on max_chunk_memory
+        # Divide the max_chunk_memory by twice the element size
+        chunk_size = self.max_chunk_memory // (element_size_bytes * 2)
+        chunk_size = min(chunk_size, num_elements)
+
+        # Split the computation into chunks
+        fp16_chunks = torch.split(fp16_output_flat, chunk_size)
+        int_w_chunks = torch.split(int_w_output_flat, chunk_size)
+
+        # Compute the loss for each chunk
+        for fp16_chunk, int_w_chunk in zip(fp16_chunks, int_w_chunks):
+            chunk_loss = (
+                (fp16_chunk.to(device) - int_w_chunk.to(device))
+                .float()
+                .pow(2)
+                .sum()
+                .item()
+            )
+            loss += chunk_loss
+
+        # Normalize the loss by the total number of elements
+        loss /= num_elements
+
+        return loss
+
+    def _set_module_kwargs(self, model, dataloader) -> None:
+        _, modules = next(iter(get_layers("re:.*layers", model).items()))
+
+        samples = [batch["input_ids"] for batch in dataloader]
+
+        samples = torch.cat(samples, dim=0)
```

Comment:
Why are you putting all samples into the same batch. This is likely the source of your memory issues

Reply:
thanks for pointing that out! will toggle this and see if it affects anything

Reply:
This was not the culprit for the OOM errors I was hitting, but this line is suspect (we are pulling this from AutoAWQ). I think I can probably get by with just
```python
samples=next(dataloader)["input_ids"]
```
I will try this before resolving this comment

Reply:
Nm, this is caching the activations for every sample in the dataloader. so only using one batch degrades the perplexity score. We need to cache all the activations, for every sample in the calibration set, so we are stuck with this being rather memory-intensive